File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/map/qualifier/HasDataSpecification.java
package org.eclipse.basyx.submodel.metamodel.map.qualifier;

import java.util.Collection;
import java.util.HashSet;
import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.qualifier.IHasDataSpecification;
import org.eclipse.basyx.submodel.metamodel.api.reference.IReference;
import org.eclipse.basyx.submodel.metamodel.map.reference.Reference;
import org.eclipse.basyx.submodel.metamodel.map.reference.ReferenceHelper;
import org.eclipse.basyx.vab.model.VABModelMap;

/**
 * HasDataSpecification class
 * 
 * @author elsheikh, schnicke
 *
 */
public class HasDataSpecification extends VABModelMap<Object> implements IHasDataSpecification {

	public static final String HASDATASPECIFICATION = "hasDataSpecification";
	/**
	 * Constructor
	 */
	public HasDataSpecification() {
		put(HASDATASPECIFICATION, new HashSet<Reference>());
	}

	public HasDataSpecification(Collection<IReference> ref) {
		put(HASDATASPECIFICATION, ref);
	}

	/**
	 * Creates a DataSpecificationIEC61360 object from a map
	 * 
	 * @param obj
	 *            a DataSpecificationIEC61360 object as raw map
	 * @return a DataSpecificationIEC61360 object, that behaves like a facade for
	 *         the given map
	 */
	public static HasDataSpecification createAsFacade(Map<String, Object> map) {
		if (map == null) {
			return null;
		}

		HasDataSpecification ret = new HasDataSpecification();
		ret.setMap(map);
		return ret;
	}

	@Override
	public Collection<IReference> getDataSpecificationReferences() {
		return ReferenceHelper.transform(get(HasDataSpecification.HASDATASPECIFICATION));
	}

	public void setDataSpecificationReferences(Collection<IReference> ref) {
		put(HasDataSpecification.HASDATASPECIFICATION, ref);
	}

}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/map/qualifier/HasDataSpecification.cpp
/*
 * HasDataSpecification.cpp
 *
 *      Author: wendel
 */

#include <BaSyx/submodel/map/qualifier/HasDataSpecification.h>

#include <BaSyx/submodel/map/reference/Reference.h>

namespace basyx {
namespace submodel {

HasDataSpecification::HasDataSpecification()
	: vab::ElementMap{}
{
	this->map.insertKey(Path::HasDataSpecification, basyx::object::make_list<basyx::object>());
}

HasDataSpecification::HasDataSpecification(basyx::object & obj)
	: vab::ElementMap{obj}
{

}

HasDataSpecification::HasDataSpecification(const basyx::specificCollection_t<IReference>& refs)
	: vab::ElementMap{}
{
	this->setDataSpecificationReferences(refs);
}

HasDataSpecification::HasDataSpecification(const IHasDataSpecification & hasDataSpecification)
  : vab::ElementMap{}
{
  this->setDataSpecificationReferences(hasDataSpecification.getDataSpecificationReferences());
}

basyx::specificCollection_t<IReference> HasDataSpecification::getDataSpecificationReferences() const
{
	auto obj_list = this->map.getProperty(Path::HasDataSpecification).Get<basyx::object::object_list_t&>();
	return vab::ElementMap::make_specific_collection<IReference, Reference>(obj_list);
}

void HasDataSpecification::setDataSpecificationReferences(const basyx::specificCollection_t<IReference> & references)
{
	auto list = basyx::object::make_list<basyx::object>();

	for (const auto & ref : references)
	{
		Reference reference{ ref->getKeys() };
		list.insert(reference.getMap());
	};

	this->map.insertKey(Path::HasDataSpecification, list, true);
}

}
}
